src/main/java/org/candlepin/model/ConsumerCurator.java

```diff
@@ -1001,6 +1008,107 @@ else if (systemUuid != null) {
         return found;
     }
 
+    /**
+     * Deletes inactive consumers and creates {@link DeletedConsumer} records.
+     * The identity certificate and content access certificate for the consumers
+     * are removed and their serials are revoked.
+     *
+     * @param inactiveConsumerIds - ids of inactive consumers to delete.
+     * @return the number of consumers that have been deleted.
+     */
+    @Transactional
+    public int deleteInactiveConsumers(Collection<String> inactiveConsumerIds) {
+        if (inactiveConsumerIds == null || inactiveConsumerIds.isEmpty()) {
+            return 0;
+        }
+
+        if (inactiveConsumerIds.size() > INACTIVE_CONSUMER_BLOCK_SIZE) {
+            String msg = "Inactive consumer ids {0} exceeds the max batch size of {1}.";
+            throw new IllegalArgumentException(
+                String.format(msg, inactiveConsumerIds.size(), INACTIVE_CONSUMER_BLOCK_SIZE));
+        }
+
+        List<Long> serialsToRevoke = new ArrayList<>(inactiveConsumerIds.size());
+        List<String> idCertsToDelete = new ArrayList<>(inactiveConsumerIds.size());
+        List<String> caCertsToDelete = new ArrayList<>(inactiveConsumerIds.size());
+
+        // Lock the consumers to prevent update requests while we are deleting.
+        Collection<Consumer> inactiveConsumers = lockAndLoad(inactiveConsumerIds);
+        inactiveConsumers.stream()
+            .filter(Objects::nonNull)
+            .forEach((consumer) -> {
+                IdentityCertificate idCert = consumer.getIdCert();
+                if (idCert != null) {
+                    idCertsToDelete.add(idCert.getId());
+                    serialsToRevoke.add(idCert.getSerial().getId());
+                }
+                ContentAccessCertificate contentAccessCert = consumer.getContentAccessCert();
+                if (contentAccessCert != null) {
+                    caCertsToDelete.add(contentAccessCert.getId());
+                    serialsToRevoke.add(contentAccessCert.getSerial().getId());
+                }
+
+                this.delete(consumer);
+            });
```

Comment:
In the design document it mentions inserting the rows into the cp_deleted_consumers table with a single insert statement with a subquery for the entire batch.  This is ideal, but I wasn't able to find a way to populate the non-null key column "id" for the cp_deleted_consumers table this way. I tried using the random_uuid function, but got an error saying I did not have permissions to use the function.

This isn't ideal as we are deleting the consumers individually and maybe doing some performance testing on this would be a good idea to get an idea on how it performs.

**Insert statement I was using:**
```
INSERT INTO cp_deleted_consumers (created, updated, consumer_uuid, owner_id, owner_displayname, owner_key, principal_name)
SELECT NOW(), NOW(), consumer.uuid, consumer.owner_id, owner.displayname, owner.account, :principalName
FROM cp_consumer consumer
JOIN cp_owner owner on owner.id=consumer.owner_id
WHERE consumer.uuid IN (:consumerIds)
```

Reply:
Couple things to note here:
- Likely due to an old database limitation or some other such decision, our UUID columns aren't *real* UUID columns; we pull the hyphens out, bringing the character count down to 32 (from 36 IIRC). So even if you had a properly functioning UUID function, it probably wouldn't fit and/or would be truncated.
- PostgreSQL and MySQL/MariaDB have different syntax for doing UUID generation, so you wouldn't be able to generalize it anyway
- HQL/JPQL don't have any nice ways of doing it directly. You could try doing the general function call (`function(...)`), but I'd have to look up how the INSERT ... INTO works with HQL/JPQL. I remember it being fairly limited, so it might not even be an option. I'm not clear on the details there atm, however.

All that said, it's not critical that we have well-formed UUIDs for our objects. It's definitely a nice-to-have, but it's not strictly required for our purposes. We've hit similar problems in other migrations and tasks, and have used the following to get around it: `MD5(CONCAT(<some safe set of unique columns and/or NOW()>))`

MD5 isn't cryptographically secure anymore, but it's still functional for non-security purposes, and it happens to generate 32-character strings. So long as the inputs are selected carefully and we do our best to minimize the chance of collision, that should be sufficient.

Reply:
Resolving this, as the new queries don't hit this issue.